### DISCUSSION TITLE
Lora use: anyone use lora?

### DISCUSSION THREAD

**Discussion by manudj195** [arendst/Tasmota#21386] [Q&A]
I'm trying to use Lora module SX1278 and ESP32-C3 Super Mini to send 2 button from Web UI or MQTT command in a place that have not WiFi/Bluetooth/Zigbee connection (Yes, Like a Bunker) but Lora is only system that still works.
I've got compile my custom Tasmota and i saw Lora setters in GPIO list, I configured it proprely but when try to configure it in console i gave "Unknow" comand
in log i saw this:
CFG: Lora use defaults as file system not ready or key not found
Which file it wants?

![IMG_8252](https://github.com/arendst/Tasmota/assets/81382442/453a0b67-5b0d-4a05-a8ca-9016a528a878)
![IMG_8253](https://github.com/arendst/Tasmota/assets/81382442/ff3ab7bd-9901-409c-9775-1325fa8177c5)

**Comment by arendst** [comment]
Pls try to provide the information asked when opening an issue. Like output of ``status 0``, used template, gpio config, boot log etc.

If GPIO is configured correctly it connects to the SX1278 and reports something like 
```
13:08:43.005 CFG: Lora loaded from file
13:08:43.029 LOR: SX127x initialized
```

This should be a valid GPIO config. Notice you cannot use every GPIO for Input/Output.

![image](https://github.com/arendst/Tasmota/assets/11044339/6d11bdc3-930e-467d-8475-df4134927d67)

**Reply by manudj195** [reply]
I configured it in this mode because the pinout of ESP32-C3 Super Mini is this

![IMG_8254](https://github.com/arendst/Tasmota/assets/81382442/8b244d83-5bbc-4b01-ae96-0095831498fd)

![IMG_8246](https://github.com/arendst/Tasmota/assets/81382442/925cf7ae-e484-40ea-9f1c-fc855dc949c3)

I also tryed 
```
GPIO 0 --> LoRa - DIO0
GPIO 1 --> LoRa - Rst
...
GPIO 4 --> SPI SCLK (1)
GPIO 5 --> SPI MISO (1)
GPIO 6 --> SPI MOSI (1)
GPIO 7 --> LoRa - CS
```
The problem is the missing file but i don't know which file
```
14:05:36.007 CFG: Lora use defaults as file system not ready or key not found
```

**Reply by arendst** [reply]
The file is made once the hardware is recognised and if you had provided the boot log I suppose the message isn't there.

As you can see from my earlier entry your GPIO config needs to use ``SPI`` instead of ``SSPI``.
I expect the C3 is picky about GPIO use. Let me reconfig mine and see if that's the case.

**Reply by sfromis** [reply]
The screenshot config is clearly bad, but the "also" looks better.

Your images shows pins on the ESP32-C3 board not being soldered. That's **really** bad.

The message you focus on is not really an error message, more like information that it did not find Lora data in the "hidden" file `.drvset000`, which is normal for initial setup. In that file, the settings are stored in JSON format under the object name `drvset73`.

**Reply by Jason2866** [reply]
I would really surprised if connecting the C3 board only via pins will work. 100% sure doing this way it is not reliable!

**Reply by arendst** [reply]
Yes. Ofcourse it's a no-go trying to make it work without soldering the pins to the module ....

**Reply by sfromis** [reply]
BTW. Even with pins soldered properly, using a breadboard is still not great for reliable connections. And worse when having the antenna face down in the middle of all those metal rails.

**Reply by fapoo** [reply]
Hi guys, I have pretty mich the same setup with an RFM95W and a C3 Super mini and I have probably a really stupid question: How do I know the exact wiring? I suppose that it is like the following:
DIO0 - LoRa Dio0
MISO - SPI MISO
MOSI - SPI MOSI
RESET - LoRa RST
SCK - SPI CLK (?)
NSS - LoRa CS (?)

At least that's what I deduced from the photos and the screenshots. But how do I know that without those?
Sorry, I am still quite clueless in this topic.

**Reply by sfromis** [reply]
Your "suppose" looks reasonable to me, but I have zero experience with that board. Still, the RFM95W page from Adafruit does not indicate a "NSS" pin, but does have "CS".
https://www.adafruit.com/product/3072

In general, you are supposed to know your hardware, typically based on whatever docs from the vendor, and match that to the gpio component names used with Tasmota. And Adafruit has a link to a tutorial for wiring.

**Comment by arendst** [comment]
Just tested this config

![image](https://github.com/arendst/Tasmota/assets/11044339/adc67cd6-79b5-4b03-ab40-1bdeeaf9a50d)

Works fine:
```
00:00:00.050 CMD: Using USB CDC
00:00:00.050 HDW: ESP32-C3 v0.4 
00:00:00.061 UFS: FlashFS mounted with 304 kB free
00:00:00.069 CFG: Loaded from File, Count 71
00:00:00.071 SER: Set to 8N1 115200 bit/s
00:00:00.071 SER: HWCDC supports 115200 bit/s only
00:00:00.075 QPC: Count 1
00:00:00.077 SPI: Hardware using GPIO04(CLK), GPIO06(MOSI) and GPIO05(MISO)
00:00:00.079 ROT: Mode 1
00:00:00.100 CFG: No '*.autoconf' file found
00:00:00.104 BRY: Berry initialized, RAM used 3712 bytes
00:00:00.109 BRY: No 'preinit.be'
00:00:00.110 SRC: Restart
00:00:00.112 Project tasmota - Esp32c3Mini1 Version 13.4.1.2(theo)-3_0_0(2024-05-08T14:36:09)
00:00:00.219 CFG: Lora loaded from file
00:00:00.241 LOR: SX127x initialized
00:00:00.255 BRY: No 'autoexec.be'
00:00:00.495 WIF: Attempting connection...
00:00:01.002 WIF: Connecting to AP1 indebuurt_IoT Channel 11 BSSId 18:E8:29:CA:17:C1 in mode HT20 as esp32c3mini1...
00:00:02.035 WIF: IPv4 192.168.2.151, mask 255.255.255.0, gateway 192.168.2.254
00:00:03.535 WIF: Connected
00:00:03.620 WIF: IPv6 Local fe80::424c:caff:fefb:4b54%st1
00:00:03.739 HTP: Web server active on esp32c3mini1 with IP address 192.168.2.151
00:00:04.946 NTP: No reply from 127.0.1.1
00:00:04.975 RTC: UTC 2024-05-08T14:17:27Z, DST 2024-03-31T02:00:00, STD 2024-10-27T03:00:00
16:17:27.001 RTC: Synced by NTP
16:17:27.756 MQT: Attempting connection...
16:17:27.772 MQT: Connected
16:17:27.775 MQT: tele/esp32c3mini1/LWT = Online (retained)
16:17:27.777 MQT: cmnd/esp32c3mini1/POWER = 
16:17:27.778 MQT: Subscribe to cmnd/esp32c3mini1/#
16:17:27.780 MQT: Subscribe to cmnd/tasmotas/#
16:17:27.781 MQT: Subscribe to cmnd/DVES_FB4B54_fb/#
16:17:27.784 MQT: tele/esp32c3mini1/INFO1 = {"Info1":{"Module":"ESP32C3","Version":"13.4.1.2(theo)","FallbackTopic":"cmnd/DVES_FB4B54_fb/","GroupTopic":"cmnd/tasmotas/"}}
16:17:27.787 MQT: tele/esp32c3mini1/INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"esp32c3mini1","IPAddress":"192.168.2.151","IP6Global":"","IP6Local":"fe80::424c:caff:fefb:4b54%st1"}}
16:17:27.791 MQT: tele/esp32c3mini1/INFO3 = {"Info3":{"RestartReason":"Software reset digital core","BootCount":38}}
16:17:29.620 WIF: IPv6 Global fd84:7996:2cbf:1349:424c:caff:fefb:4b54
16:17:29.915 QPC: Reset
16:17:30.213 HTP: Main Menu
16:17:31.910 APP: Boot Count 38
16:17:31.916 MQT: tele/esp32c3mini1/STATE = {"Time":"2024-05-08T16:17:31","Uptime":"0T00:00:09","UptimeSec":9,"Heap":204,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Berry":{"HeapUsed":3,"Objects":34},"Wifi":{"AP":1,"SSId":"indebuurt_IoT","BSSId":"18:E8:29:CA:17:C1","Channel":11,"Mode":"HT20","RSSI":100,"Signal":-46,"LinkCount":1,"Downtime":"0T00:00:03"}}
16:17:32.333 CFG: Saved, Count 72, Bytes 4096
16:17:33.939 MQT: tele/esp32c3mini1/RESULT = {"LoRaReceived":"40E464C400002A000A6DEF0C398E64FA81F968C86E8FCE","RSSI":-46.0,"SNR":13.5}
16:17:37.004 MQT: tele/esp32c3mini1/RESULT = {"LoRaReceived":"40E464C400002B000AC2C92867380DF510728EAAD24253","RSSI":-45.0,"SNR":10.0}
16:17:43.825 HTP: Configuration
16:17:45.009 HTP: Configure Module
```

**Comment by wili98w** [comment]
Im waiting for such a bridge for a longtime and finally seems it is here and I've been able to configure Tasmota and seems like it listen for any traffic.
Since my idea is to send Lora (not LoraWAN) messages, I do not understand what shall be the message format to be processed by Tasmota.
Im willing to utilize Atmega 328P with limited resources, therefore Im willing to avoid LoraWAN and use Lora only, 
Can somebody point a link or source code where I can find the message format I've shall generate in order to be received by Tasmota standard Lora bridge?

**Reply by sfromis** [reply]
Tasmota does not stipulate any format for messages, just send anything in character format, only mild deviation is if allowing reception of LoRaCommand messages (no security), but with the message starting with the topic of the receiver device.
You do not have to limit yourself to character data, see
https://tasmota.github.io/docs/LoRa-and-LoRaWan-Bridge/#lora-commands

Having two Tasmota devices with operational LoRa, I can just do a command like: `LoRaSend test`, and see the message on the other device.

**Reply by nikito7** [reply]
You need to use same lora parameters. Syncword etc
